src/lib.rs

```diff
@@ -258,96 +260,135 @@ impl Parser {
     pub fn send_text(&mut self, text: &str) -> events::TelnetEvents {
         events::TelnetEvents::build_send(Parser::escape_iac(format!("{}\r\n", text).into_bytes()))
     }
-    /// The internal parser method that takes the current buffer and generates the corresponding events.
-    fn process(&mut self) -> Vec<events::TelnetEvents> {
-        let mut t: Vec<Vec<u8>> = Vec::with_capacity(4);
-        let mut event_list: Vec<events::TelnetEvents> = Vec::with_capacity(2);
+
+    /// Extract sub-buffers from the current buffer
+    fn extract_event_data(&mut self) -> Vec<Vec<u8>> {
+        enum State {
+            Normal,
+            IAC,
+            Neg,
+            Sub,
+        };
+        let mut iter_state = State::Normal;
+
+        let mut events: Vec<Vec<u8>> = Vec::with_capacity(4);
         let iter = self.buffer.iter().enumerate();
-        let mut offset_next: usize;
-        let mut offset_last: usize = 0;
+        let mut cmd_begin: usize = 0;
+
         for (index, &val) in iter {
-            if val == 255 && self.buffer[index + 1] != 255 {
-                offset_next = index;
-                if offset_next != offset_last {
-                    if self.buffer[offset_last + 1] == 250 && self.buffer[offset_next + 1] == 240 {
-                        offset_next += 2;
+            match iter_state {
+                State::Normal => {
+                    if val == IAC {
+                        if cmd_begin < index {
+                            events.push(Vec::from(&self.buffer[cmd_begin..index]));
+                        }
+                        cmd_begin = index;
+                        iter_state = State::IAC;
                     }
-                    t.push(Vec::from(&self.buffer[offset_last..offset_next]));
-                    offset_last = offset_next;
-                }
+                },
+                State::IAC => {
+                    match val {
+                        IAC => iter_state = State::Normal, // Double IAC, ignore
+                        GA | NOP => {
+                            events.push(Vec::from(&self.buffer[cmd_begin..index+1]));
+                            cmd_begin = index + 1;
+                            iter_state = State::Normal;
+                        },
+                        SB => iter_state = State::Sub,
+                        _ => iter_state = State::Neg, // WILL | WONT | DO | DONT | IS | SEND 
```

Comment:
Not sure if these are the only IAC negotations that have 3 bytes. That's what I gathered from reading the protocol but it's a pretty heavy read so I might have missed something.

Reply:
As far as I am aware they are the only ones that have 3 bytes.